arch/x86/kvm/vmx/pkvm/hyp/iommu.c

```diff
@@ -321,41 +321,33 @@ static bool sync_shadow_context_entry(struct pgt_sync_data *sdata)
 	struct context_entry *shadow_ce = sdata->shadow_ptep, tmp = {0};
 	struct context_entry *guest_ce = sdata->guest_ptep;
 	bool updated = false;
+	u8 aw;
 
-	/*
-	 * If hardware PT translation(ecap(bit6)) is being used by host,
-	 * pKVM won't shadow I/O paging structures as there are no virtual
-	 * I/O paging structures. Instead, pKVM should consider using EPT
-	 * as the 2nd-stage I/O table to ensure the memory protection.
-	 * (i.e. In legacy mode, although the virtual translation type is
-	 * CONTEXT_TT_PASS_THROUGH, the real translation type is
-	 * CONTEXT_TT_MULTI_LEVEL in pKVM.)
-	 *
-	 * Todo: to fully support legacy mode vIOMMU, either by using EPT as
-	 * 2nd-stage I/O translation table when HW PT translation is being
-	 * used by host or by disabling the HW PT translation cap so that
-	 * virtual I/O paging structures will come out which can be shadowed
-	 * by pKVM IOMMU for the DMA restriction.
-	 *
-	 * (Note: Currently we only support shadow root/context table in legacy
-	 * mode. The shadowing paging structures hasn't been supported yet.
-	 * The the following logic added for LM is to workaround the case in which
-	 * the type is set as CONTEXT_TT_PASS_THROUGH in host. W/o this workaround,
-	 * when virtual vt-d sets a domain type to PASS_THROUGH, the sdata->shadow_pa
-	 * would be 0 as there is no virtual table for this domain in host. However
-	 * "sdata->shadow_pa == 0" would lead shadow context entry not to be updated.
-	 * So add a workaround to make the shadow context entry be updated anyway.
-	 * After the missing piece of shadowing paging structures of legacy mode is
-	 * added, we won't need this workaround anymore.)
-	 */
-	if (sdata->guest_ptep && (sdata->shadow_pa ||
-		(!ecap_smts(sdata->iommu_ecap) &&
-		 (context_lm_get_tt(guest_ce) == CONTEXT_TT_PASS_THROUGH)))) {
+	if (sdata->guest_ptep && sdata->shadow_pa) {
 		tmp.hi = guest_ce->hi;
 		tmp.lo = sdata->shadow_pa | (guest_ce->lo & 0xfff);
 
-		/* Clear DTE to make sure device TLB is disabled for security */
-		context_clear_dte(&tmp);
+		if (ecap_smts(sdata->iommu_ecap))
+			/* Clear DTE to make sure device TLB is disabled for security */
+			context_clear_dte(&tmp);
```

Comment:
thanks for fixing this potential issue, maybe we should change the function name to context_sm_clear_dte().